source/extensions/filters/http/common/aws/signer_impl.cc

```diff
@@ -0,0 +1,256 @@
+          lines.emplace_back(StringUtil::trim(line));
+        }
+        auto value = StringUtil::join(lines, ",");
+        // Remove duplicate spaces
+        const auto end = std::unique(value.begin(), value.end(), [](char lhs, char rhs) {
+          return (lhs == rhs) && (lhs == ' ');
+        });
+        value.erase(end, value.end());
+        map->emplace(entry.key().c_str(), value);
+        return Http::HeaderMap::Iterate::Continue;
+      },
+      &out);
+  // The AWS SDK has a quirk where it removes "default ports" (80, 443) from the host headers
+  // Additionally, we canonicalize the :authority header as "host"
+  const auto* authority_header = headers.Host();
+  if (authority_header != nullptr && !authority_header->value().empty()) {
+    const auto& value = authority_header->value().getStringView();
+    const auto parts = StringUtil::splitToken(value, ":");
+    if (parts.size() > 1 && (parts[1] == "80" || parts[1] == "443")) {
+      out.emplace(Http::Headers::get().HostLegacy.get(),
+                  std::string(parts[0].data(), parts[0].size()));
+    } else {
+      out.emplace(Http::Headers::get().HostLegacy.get(), std::string(value.data(), value.size()));
+    }
+  }
+  return out;
+}
+
+std::vector<uint8_t> SignerImpl::hash(const Buffer::Instance& buffer) const {
+  std::vector<uint8_t> digest(SHA256_DIGEST_LENGTH);
+  EVP_MD_CTX ctx;
+  auto code = EVP_DigestInit(&ctx, EVP_sha256());
+  RELEASE_ASSERT(code == 1, "Failed to init digest context");
+  const auto num_slices = buffer.getRawSlices(nullptr, 0);
+  STACK_ARRAY(slices, Buffer::RawSlice, num_slices);
+  buffer.getRawSlices(slices.begin(), num_slices);
+  for (const auto& slice : slices) {
+    code = EVP_DigestUpdate(&ctx, slice.mem_, slice.len_);
+    RELEASE_ASSERT(code == 1, "Failed to update digest");
+  }
+  unsigned int digest_length;
+  code = EVP_DigestFinal(&ctx, digest.data(), &digest_length);
+  RELEASE_ASSERT(code == 1, "Failed to finalize digest");
+  RELEASE_ASSERT(digest_length == SHA256_DIGEST_LENGTH, "Digest length mismatch");
+  return digest;
+}
+
+std::vector<uint8_t> SignerImpl::hmac(const std::vector<uint8_t>& key,
+                                      const std::string& string) const {
+  std::vector<uint8_t> mac(EVP_MAX_MD_SIZE);
+  HMAC_CTX ctx;
+  RELEASE_ASSERT(key.size() < std::numeric_limits<int>::max(), "Hmac key is too long");
+  HMAC_CTX_init(&ctx);
+  auto code = HMAC_Init_ex(&ctx, key.data(), static_cast<int>(key.size()), EVP_sha256(), nullptr);
+  RELEASE_ASSERT(code == 1, "Failed to init hmac context");
+  code = HMAC_Update(&ctx, reinterpret_cast<const uint8_t*>(string.data()), string.size());
+  RELEASE_ASSERT(code == 1, "Failed to update hmac");
+  unsigned int len;
+  code = HMAC_Final(&ctx, mac.data(), &len);
+  RELEASE_ASSERT(code == 1, "Failed to finalize hmac");
+  RELEASE_ASSERT(len <= EVP_MAX_MD_SIZE, "Hmac length too large");
+  HMAC_CTX_cleanup(&ctx);
+  mac.resize(len);
+  return mac;
```

Comment:
I think hmac and hash are general enough to put in `source/common/ssl/utility.cc` or somewhere like that, this seems like something other folks might want to reuse.

Reply:
Thats a good idea

Reply:
FYI, @bdecoste is working on moving SSL library dependent stuff to `extensions`, to allow compiling out BoringSSL #5543 , I think this might be fine but not sure how SSL library specific this is (BoringSSL vs OpenSSL).